### PR TITLE
cli: add cli completions

### PR DESCRIPTION
**Pull request by ad-tra**
fixes: #878

### CODE REVIEW

**Comment by ad-tra** [comment]
#### BASH 

Append to  `~/.bashrc`:

`source < {path_to_lb_completion_script}`


#### ZSH 

Append the following to `~/.zshrc`:

    autoload -U +X bashcompinit && bashcompinit
    source < {path_to_lb_completion_script}

**Review by Parth — CHANGES_REQUESTED** [review]

File: clients/cli/src/completions.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    match shell {
+        Shell::Bash | Shell::Zsh => print!("{}", BASH_COMPLETIONS),
+        Shell::Fish => todo!(),
+        _ => todo!(),
+    }
+
+    Ok(())
+}
+
+pub fn complete(core: &Core, input: String, current: i32) -> Result<(), CliError> {
```
[the remark this answers is not in the document]
Comment (ad-tra) [reply clients/cli/src/completions.rs:47]: i converted `CliError` into an enum that houses two error variants, one that prints to console and one that doesn't (it prints to console on non release builds for ease of development)

File: clients/cli/src/completions.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    LbFilePath,
+    LbAnyPath,
+    LbFolderPath,
+    PendingShareId,
+    None,
+}
+
+pub fn generate_completions(shell: Shell) -> Result<(), CliError> {
+    match shell {
+        Shell::Bash | Shell::Zsh => print!("{}", BASH_COMPLETIONS),
```
[the remark this answers is not in the document]
Comment (ad-tra) [reply clients/cli/src/completions.rs:39]: to install bash completions on zsh, you need to [init them](https://stackoverflow.com/questions/67136714/how-to-properly-call-compinit-and-bashcompinit-in-zsh) with `autoload -U +X bashcompinit && bashcompinit` did you do that?

File: clients/cli/src/completions.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        .ok_or(CliError("no subcommand provided in input".to_string()))?;
+
+    let binding = LbCli::command();
+
+    //the argument to provide completions for
+    let selected_arg = binding
+        .find_subcommand(matched_subcommand.0)
+        .ok_or(CliError("subcommand undefined".to_string()))?
+        .get_arguments()
+        .nth(current as usize - 2)
```
[the remark this answers is not in the document]
Comment (ad-tra) [reply clients/cli/src/completions.rs:68]: the input array: `[lockbook, subcommand (ex: edit), arg 1, arg2, .. , arg_n ]`
the arguments array: `[arg 1, arg2, .. , arg_n ]`

we want the current arg from the args array, based on the cursor in the input array. hence the offset of 2

the case where there will be index out of bounds is if you don't provide a subcommand. but this arg finding segment wouldn't execute as the no subcommand error case is caught earlier by  

```rust 

    let matched_subcommand = matches
        .subcommand()
        .ok_or(CliError::SilentError("no subcommand provided in the input".to_string()))?;
```

File: clients/cli/src/error.rs [hunk 4]
```diff
@@ -1,34 +1,52 @@
...
 use std::io;
 
-pub struct CliError(pub String);
-
-impl CliError {
-    pub fn new(msg: impl ToString) -> Self {
-        Self(msg.to_string())
-    }
+pub enum CliError {
+    ConsoleError(String),
```
Comment (Parth) [inline clients/cli/src/error.rs:5]:
```suggestion
    Console(String),
    Silent(String),
```

**Review by Parth — APPROVED** [review]
LGTM, a lil unsure about how we are presently handling zsh bash compatibility, I think we ought to make it more clear that we're falling back to a compat mode so someone knows to config it

**Pull request merged**